File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        Stack<TreeNode> st = new Stack<>();    
        st.push(root);
        while(!st.empty()){
            while(root!=null && root.left!=null){
                st.push(root.left);
                root = root.left;
            }
            root = st.pop();
            k--;
            if(k==0) break;
            if(root.right != null){                
                st.push(root.right);                
            }                       
            root = root.right;  
        }
        return root.val;
    }
}

File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        st = []   
        st.append(root)    
        while st:
            top = st.pop()
            while top:
                st.append(top)
                top = top.left
            k -= 1
            curr = st.pop()
            if k == 0:
                return curr.val             
            if curr.right:
                st.append(curr.right)
            else:
                st.append(None)